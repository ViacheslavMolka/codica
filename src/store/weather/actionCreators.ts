import {
  GetWeatherAction,
  GetWeatherErrorAction,
  HourlyGetWeatherAction,
  UpdateGetWeatherAction,
  WeatherActionEnum
} from './actionTypes';

export const loadWeather = (payload: Array<object>): GetWeatherAction => ({ type: WeatherActionEnum.GET_WEATHER, payload });

export const newWeather = (payload: Array<object>): UpdateGetWeatherAction => ({ type: WeatherActionEnum.UPDATE_WEATHER, payload });

export const getHourlyWeather = (payload: Array<object>): HourlyGetWeatherAction => ({ type: WeatherActionEnum.HOURLY_WEATHER, payload });

export const getWeatherError = (payload: string): GetWeatherErrorAction => ({ type: WeatherActionEnum.ERROR, payload });
