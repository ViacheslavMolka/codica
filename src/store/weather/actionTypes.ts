export enum WeatherActionEnum {
  GET_WEATHER = "GET_WEATHER",
  UPDATE_WEATHER = "UPDATE_WEATHER",
  HOURLY_WEATHER = "HOURLY_WEATHER",
  ERROR = "ERROR",
}

export interface WeatherState {
  weather: Array<object>,
	hourly: Array<object>,
	error: string;
}

export interface GetWeatherAction {
  type: WeatherActionEnum.GET_WEATHER;
  payload: any;
}

export interface UpdateGetWeatherAction {
  type: WeatherActionEnum.UPDATE_WEATHER;
  payload: any;
}

export interface HourlyGetWeatherAction {
  type: WeatherActionEnum.HOURLY_WEATHER;
  payload: Array<object>;
}

export interface GetWeatherErrorAction {
  type: WeatherActionEnum.ERROR;
  payload: string;
}

export type WeatherActions =
  GetWeatherAction |
  UpdateGetWeatherAction |
  HourlyGetWeatherAction |
  GetWeatherErrorAction
