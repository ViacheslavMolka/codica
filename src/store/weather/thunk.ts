import { AppDispatch } from "..";
import {
  loadWeather,
  newWeather,
  getHourlyWeather,
  getWeatherError,
} from "./actionCreators";

export const getWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(loadWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};

export const updateWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(newWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};

export const hourlyWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/forecast?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(getHourlyWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};
