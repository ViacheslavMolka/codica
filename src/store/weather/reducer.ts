/* eslint-disable @typescript-eslint/no-explicit-any */
import { WeatherActionEnum, WeatherState, WeatherActions, IWeather } from './actionTypes';

const initialState: WeatherState = {
	weather: [],
	hourly: [],
	error: '',
};

export default function weather(state = initialState, action: WeatherActions): WeatherState {
	switch (action.type) {
		case WeatherActionEnum.GET_WEATHER: {
			const isHas: IWeather | any = state.weather.find((item: any) => item.name === action.payload.name);
			const checkWeather = isHas ? [] : [action.payload];
      return {
				...state,
        weather: [...state.weather, ...checkWeather],
      }}
		case WeatherActionEnum.UPDATE_WEATHER: {
			const newWeather: Array<IWeather> = state.weather.map((item: any) => {
				if (item.name === action.payload.name) {
					return action.payload
				}
				return item;
			})
			return {
				...state,
				weather: newWeather,
			}
		}
		case WeatherActionEnum.HOURLY_WEATHER:
      return {
				...state,
        hourly: [...state.hourly, action.payload],
      }
		case WeatherActionEnum.ERROR:
			return {
				...state,
				error: action.payload,
			}
		default:
			return state;
	}
}
