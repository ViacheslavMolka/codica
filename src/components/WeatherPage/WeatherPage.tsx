/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { Box } from '@mui/system';

import WeatherCard from '../WeatherCard/WeatherCard';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import { getWeather } from '../../store/weather/thunk';
import { useAppDispatch } from '../../store';
import { Button, TextField } from '@mui/material';

const WeatherPage: React.FC = () => {
  const { weather = [] } = useTypedSelector(state => state.weather);
  const [city, setCity] = React.useState('');
  const dispatch = useAppDispatch();

  React.useEffect(() => {
    !!city && localStorage.setItem(`${city}`, city)
  }, [weather])

  const handleSubmit = () => {
    dispatch(getWeather(city));
  };
  return (
    <Box>
      <Box sx={{ display: 'flex', justifyContent: 'center' }}>
        <TextField 
          id="city"
          placeholder="Type city"
          variant="outlined"
          value={city}
          onChange={e => setCity(e.target.value)}
        />
        <Button sx={{ marginLeft: 2 }} onClick={handleSubmit} variant="outlined" disabled={!city}>
          Add sity
        </Button>
      </Box>
      <Box sx={{ display: 'flex', flexWrap: 'wrap' }}>
        {weather.map((item: any) => (
          <WeatherCard key={item.name} info={item} />
        ))}
      </Box>
    </Box>
  );
};

export default WeatherPage;
