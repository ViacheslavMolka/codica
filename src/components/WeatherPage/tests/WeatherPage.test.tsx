import * as React from 'react';
import * as reactRedux from 'react-redux';
import renderer from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';

import weather from '../../../store/weather/reducer';
import WeatherPage from '../WeatherPage';


jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('components > WeatherPage component', () => {
  const useSelectorMock = reactRedux.useSelector;
  const store = createStore(combineReducers({ weather }));
  const useTypedSelector: any = useSelectorMock;
  const mockStore = {
    hourly: [{ name: 'Lviv' }],
    weather: [],
    error: '',
  };
  beforeEach(() => {
    useTypedSelector.mockImplementation((selector: any) => selector(mockStore));
  })
  afterEach(() => {
    useTypedSelector.mockClear();
  })

  it('should render WeatherPage component', () => {
    const renderedComponnet = renderer.create(
      <Provider store={store}>
        <WeatherPage />
      </Provider>
    );
    expect(renderedComponnet).toMatchSnapshot('should render correctly');
  });
});

