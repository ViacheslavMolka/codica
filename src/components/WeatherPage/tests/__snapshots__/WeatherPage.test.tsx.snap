// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > WeatherPage component should render WeatherPage component: should render correctly 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="MuiBox-root css-1l4w6pd"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="city"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Type city"
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-hdw1oc"
          >
            <span
              className="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root Mui-disabled  css-1y7fz8o-MuiButtonBase-root-MuiButton-root"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Add sity
    </button>
  </div>
  <div
    className="MuiBox-root css-2imjyh"
  />
</div>
`;
